src/igraph/__init__.py

```diff
@@ -3211,6 +3211,56 @@ def get_incidence(self, types="type", *args, **kwds):
         """
         return super(Graph, self).get_incidence(types, *args, **kwds)
 
+    ###########################
+    # DFS (C version will come soon)
+    def dfs(self, vid, mode=OUT):
+        """Conducts a depth first search (DFS) on the graph.
+
+        @param vid: the root vertex ID
+        @param mode: either L{IN} or L{OUT} or L{ALL}, ignored
+          for undirected graphs.
+        @return: a tuple with the following items:
+           - The vertex IDs visited (in order)
+           - The parent of every vertex in the DFS
+        """
+        nv = self.vcount()
+        added = [False for v in range(nv)]
+        stack = []
+
+        # prepare output
+        vids = []
+        parents = []
+
+        # ok start from vid
+        stack.append(vid)
+        vids.append(vid)
+        parents.append(vid)
+        added[vid] = True
+
+        # go down the rabbit hole
+        while stack:
+            vid = stack[-1]
+            if mode == IN:
+                edges = self.vs[vid].in_edges()
```

Comment:
Can't we simply use `self.neighbors(vid, mode=mode)` here? This would avoid creating a `Vertex` object (the output of `self.vs[vid]`) and would also simplify the code that comes later when we have to figure out which is the "other" endpoint of the edge.

Reply:
Another trick that comes into my mind: we could have another stack where we push the retrieved neighbors of `vid` when it is visited for the first time. Then, whenever we need to find the _next_ neighbor of `vid` to visit, we simply pop off the last element of the list. If the list is empty, there are no more neighbors of `vid` to visit so we can backtrack.

Reply:
Alright I use neighbors now, good catch.

I tried playing a little with a different stack with *all* new neighbors, but it didn't make much of a change in terms of speed. I guess we could use a heap to keep the things sorted by distance and parent, but I'm not sure it improves a lot considering it's pure Python anyway.